File: inputpanel/inputpanel.go
package inputpanel

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	tea "charm.land/bubbletea/v2"
	"charm.land/bubbles/v2/textinput"
	"charm.land/lipgloss/v2"
)

// Result messages sent to parent
type SubmitMsg struct{ Value string }
type CancelMsg struct{}

type Mode int

const (
	ModeText Mode = iota // Simple text input (rename, project name)
	ModePath             // Path input with directory listing
)

type dirEntry struct {
	name  string
	isDir bool
}

type Model struct {
	input      textinput.Model
	mode       Mode
	width      int
	height     int
	active     bool
	title      string

	// Directory listing (for ModePath)
	dirEntries []dirEntry
	dirCursor  int
	dirScroll  int
}

func New() Model {
	ti := textinput.New()
	ti.Focus()
	ti.CharLimit = 256

	return Model{
		input: ti,
	}
}

func (m *Model) Activate(mode Mode, title string, initialValue string, width, height int) {
	m.mode = mode
	m.title = title
	m.width = width
	m.height = height
	m.active = true
	m.input.SetValue(initialValue)
	m.input.CursorEnd()
	m.input.Focus()
	m.dirCursor = 0
	m.dirScroll = 0
	if mode == ModePath {
		m.updateDirListing()
	}
}

func (m *Model) Deactivate() {
	m.active = false
	m.input.Blur()
}

func (m Model) Active() bool {
	return m.active
}

func (m Model) Value() string {
	return m.input.Value()
}

func (m Model) Init() tea.Cmd {
	return textinput.Blink
}

func (m Model) Update(msg tea.Msg) (Model, tea.Cmd) {
	if !m.active {
		return m, nil
	}

	switch msg := msg.(type) {
	case tea.KeyMsg:
		switch msg.String() {
		case "esc", "escape":
			m.Deactivate()
			return m, func() tea.Msg { return CancelMsg{} }

		case "enter":
			if m.mode == ModePath && len(m.dirEntries) > 0 {
				// Enter on a directory entry descends into it
				entry := m.dirEntries[m.dirCursor]
				if entry.isDir {
					current := m.expandPath(m.input.Value())
					dir := current
					prefix := ""
					if !strings.HasSuffix(current, "/") {
						dir = filepath.Dir(current)
						prefix = ""
						_ = prefix
					}
					newPath := filepath.Join(dir, entry.name) + "/"
					m.input.SetValue(newPath)
					m.input.CursorEnd()
					m.dirCursor = 0
					m.dirScroll = 0
					m.updateDirListing()
					return m, nil
				}
			}
			val := m.input.Value()
			m.Deactivate()
			return m, func() tea.Msg { return SubmitMsg{Value: val} }

		case "tab":
			if m.mode == ModePath {
				m.tabComplete()
				return m, nil
			}

		case "up":
			if m.mode == ModePath && m.dirCursor > 0 {
				m.dirCursor--
				if m.dirCursor < m.dirScroll {
					m.dirScroll = m.dirCursor
				}
				return m, nil
			}

		case "down":
			if m.mode == ModePath && m.dirCursor < len(m.dirEntries)-1 {
				m.dirCursor++
				maxVisible := m.maxVisibleEntries()
				if m.dirCursor >= m.dirScroll+maxVisible {
					m.dirScroll = m.dirCursor - maxVisible + 1
				}
				return m, nil
			}
		}
	}

	// Pass to textinput for character handling, cursor movement, etc.
	var cmd tea.Cmd
	m.input, cmd = m.input.Update(msg)

	// Update directory listing when text changes
	if m.mode == ModePath {
		m.updateDirListing()
	}

	return m, cmd
}

func (m Model) View() string {
	if !m.active {
		return ""
	}

	dimStyle := lipgloss.NewStyle().Faint(true)
	titleStyle := lipgloss.NewStyle().Bold(true).Foreground(lipgloss.Color("#51afef"))
	selectedStyle := lipgloss.NewStyle().Bold(true).Reverse(true)
	dirStyle := lipgloss.NewStyle().Foreground(lipgloss.Color("#51afef"))

	var b strings.Builder

	// Top border
	b.WriteString(dimStyle.Render(strings.Repeat("─", m.width)) + "\n")

	// Title + input
	b.WriteString(titleStyle.Render(m.title) + "\n")
	b.WriteString(m.input.View() + "\n")

	if m.mode == ModePath {
		b.WriteString(dimStyle.Render(strings.Repeat("─", m.width)) + "\n")

		maxVisible := m.maxVisibleEntries()
		end := m.dirScroll + maxVisible
		if end > len(m.dirEntries) {
			end = len(m.dirEntries)
		}

		for i := m.dirScroll; i < end; i++ {
			entry := m.dirEntries[i]
			name := entry.name
			suffix := ""
			style := dimStyle
			if entry.isDir {
				suffix = "/"
				style = dirStyle
			}

			line := fmt.Sprintf("  %s%s", name, suffix)
			if i == m.dirCursor {
				line = selectedStyle.Render(line)
			} else {
				line = style.Render(line)
			}
			b.WriteString(line + "\n")
		}

		if len(m.dirEntries) == 0 {
			b.WriteString(dimStyle.Render("  (empty)") + "\n")
		}
	}

	// Help line
	help := " esc cancel"
	if m.mode == ModePath {
		help = " esc cancel  tab complete  ↑↓ select  enter open/confirm"
	}
	b.WriteString(dimStyle.Render(help))

	return b.String()
}

// --- Internal helpers ---

func (m *Model) updateDirListing() {
	path := m.expandPath(m.input.Value())

	dir := path
	prefix := ""
	if !strings.HasSuffix(path, "/") {
		dir = filepath.Dir(path)
		prefix = strings.ToLower(filepath.Base(path))
	}

	entries, err := os.ReadDir(dir)
	if err != nil {
		m.dirEntries = nil
		return
	}

	m.dirEntries = nil
	for _, e := range entries {
		name := e.Name()
		if strings.HasPrefix(name, ".") {
			continue
		}
		if prefix != "" && !strings.HasPrefix(strings.ToLower(name), prefix) {
			continue
		}
		m.dirEntries = append(m.dirEntries, dirEntry{name: name, isDir: e.IsDir()})
	}

	if m.dirCursor >= len(m.dirEntries) {
		m.dirCursor = len(m.dirEntries) - 1
	}
	if m.dirCursor < 0 {
		m.dirCursor = 0
	}
}

func (m *Model) tabComplete() {
	path := m.expandPath(m.input.Value())

	dir := path
	prefix := ""
	if !strings.HasSuffix(path, "/") {
		dir = filepath.Dir(path)
		prefix = filepath.Base(path)
	}

	entries, err := os.ReadDir(dir)
	if err != nil {
		return
	}

	var matches []string
	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		name := e.Name()
		if strings.HasPrefix(strings.ToLower(name), strings.ToLower(prefix)) {
			matches = append(matches, name)
		}
	}

	if len(matches) == 1 {
		m.input.SetValue(filepath.Join(dir, matches[0]) + "/")
		m.input.CursorEnd()
	} else if len(matches) > 1 {
		common := matches[0]
		for _, match := range matches[1:] {
			for i := range common {
				if i >= len(match) || common[i] != match[i] {
					common = common[:i]
					break
				}
			}
		}
		if len(common) > len(prefix) {
			m.input.SetValue(filepath.Join(dir, common))
			m.input.CursorEnd()
		}
	}

	m.dirCursor = 0
	m.dirScroll = 0
	m.updateDirListing()
}

func (m Model) expandPath(path string) string {
	if strings.HasPrefix(path, "~") {
		home, _ := os.UserHomeDir()
		path = home + path[1:]
	}
	return path
}

func (m Model) maxVisibleEntries() int {
	// Reserve lines for: border, title, input, separator, help
	available := m.height - 5
	if available < 3 {
		available = 3
	}
	return available
}

File: src/input_panel.rs
use crossterm::event::{KeyCode, KeyEvent};
use ratatui::buffer::Buffer;
use ratatui::layout::Rect;
use ratatui::style::{Color, Modifier, Style};
use ratatui::text::{Line, Span};
use std::fs;

#[derive(Clone, Copy, PartialEq)]
pub enum InputMode {
    Text,
    Path,
}

pub enum InputAction {
    None,
    Submit(String),
    Cancel,
}

#[derive(Clone, Copy, PartialEq)]
enum DirEntryKind {
    /// The pinned "Select folder" row that submits the current path verbatim.
    Sentinel,
    Dir,
    File,
}

struct DirEntry {
    name: String,
    kind: DirEntryKind,
}

pub struct InputPanel {
    mode: InputMode,
    title: String,
    value: String,
    cursor_pos: usize,
    active: bool,
    width: u16,
    height: u16,
    dir_entries: Vec<DirEntry>,
    dir_cursor: usize,
    dir_scroll: usize,
}

impl InputPanel {
    pub fn new() -> Self {
        InputPanel {
            mode: InputMode::Text,
            title: String::new(),
            value: String::new(),
            cursor_pos: 0,
            active: false,
            width: 80,
            height: 15,
            dir_entries: Vec::new(),
            dir_cursor: 0,
            dir_scroll: 0,
        }
    }

    pub fn activate(
        &mut self,
        mode: InputMode,
        title: &str,
        initial_value: &str,
        width: u16,
        height: u16,
    ) {
        self.mode = mode;
        self.title = title.to_string();
        self.value = initial_value.to_string();
        self.cursor_pos = self.value.len();
        self.active = true;
        self.width = width;
        self.height = height;
        self.dir_cursor = 0;
        self.dir_scroll = 0;
        if mode == InputMode::Path {
            self.update_dir_listing();
            // The sentinel "Select folder" row sits at index 0; when there's any
            // real entry, start the cursor on it instead so reflexive Enter
            // doesn't submit the initial path before the user has navigated.
            if self.dir_entries.len() > 1 {
                self.dir_cursor = 1;
            }
        }
    }

    pub fn deactivate(&mut self) {
        self.active = false;
    }

    pub fn is_active(&self) -> bool {
        self.active
    }

    pub fn cursor_position(&self, area: Rect) -> (u16, u16) {
        // Title is on line 1 (after border), input on line 2
        let x = area.x + self.cursor_pos as u16;
        let y = area.y + 2;
        (x, y)
    }

    pub fn handle_key(&mut self, key: &KeyEvent) -> InputAction {
        if !self.active {
            return InputAction::None;
        }

        match key.code {
            KeyCode::Esc => {
                self.deactivate();
                InputAction::Cancel
            }

            KeyCode::Enter => {
                if self.mode == InputMode::Path && !self.dir_entries.is_empty() {
                    let entry = &self.dir_entries[self.dir_cursor];
                    match entry.kind {
                        DirEntryKind::Sentinel => {
                            let val = self.expand_path(&self.value);
                            self.deactivate();
                            return InputAction::Submit(val);
                        }
                        DirEntryKind::Dir => {
                            let current = self.expand_path(&self.value);
                            let dir = if current.ends_with('/') {
                                current
                            } else {
                                match current.rfind('/') {
                                    Some(i) => current[..=i].to_string(),
                                    None => current,
                                }
                            };
                            let new_path = format!("{}{}/", dir, entry.name);
                            self.value = new_path;
                            self.cursor_pos = self.value.len();
                            self.dir_cursor = 0;
                            self.dir_scroll = 0;
                            self.update_dir_listing();
                            if self.dir_entries.len() > 1 {
                                self.dir_cursor = 1;
                            }
                            return InputAction::None;
                        }
                        DirEntryKind::File => {
                            // Fall through to the generic submit below.
                        }
                    }
                }
                let val = self.value.clone();
                self.deactivate();
                InputAction::Submit(val)
            }

            KeyCode::Tab => {
                if self.mode == InputMode::Path {
                    self.tab_complete();
                }
                InputAction::None
            }

            KeyCode::Up => {
                if self.mode == InputMode::Path && self.dir_cursor > 0 {
                    self.dir_cursor -= 1;
                    if self.dir_cursor < self.dir_scroll {
                        self.dir_scroll = self.dir_cursor;
                    }
                }
                InputAction::None
            }

            KeyCode::Down => {
                if self.mode == InputMode::Path
                    && self.dir_cursor < self.dir_entries.len().saturating_sub(1)
                {
                    self.dir_cursor += 1;
                    let max_visible = self.max_visible_entries();
                    if self.dir_cursor >= self.dir_scroll + max_visible {
                        self.dir_scroll = self.dir_cursor - max_visible + 1;
                    }
                }
                InputAction::None
            }

            KeyCode::Char(c) => {
                self.value.insert(self.cursor_pos, c);
                self.cursor_pos += c.len_utf8();
                if self.mode == InputMode::Path {
                    self.update_dir_listing();
                }
                InputAction::None
            }

            KeyCode::Backspace => {
                if self.cursor_pos > 0 {
                    // Find the previous char boundary
                    let prev = self.value[..self.cursor_pos]
                        .char_indices()
                        .last()
                        .map(|(i, _)| i)
                        .unwrap_or(0);
                    self.value.remove(prev);
                    self.cursor_pos = prev;
                    if self.mode == InputMode::Path {
                        self.update_dir_listing();
                    }
                }
                InputAction::None
            }

            KeyCode::Left => {
                if self.cursor_pos > 0 {
                    self.cursor_pos = self.value[..self.cursor_pos]
                        .char_indices()
                        .last()
                        .map(|(i, _)| i)
                        .unwrap_or(0);
                }
                InputAction::None
            }

            KeyCode::Right => {
                if self.cursor_pos < self.value.len() {
                    self.cursor_pos += self.value[self.cursor_pos..]
                        .chars()
                        .next()
                        .map(|c| c.len_utf8())
                        .unwrap_or(0);
                }
                InputAction::None
            }

            _ => InputAction::None,
        }
    }

    pub fn render(&self, area: Rect, buf: &mut Buffer) {
        if !self.active {
            return;
        }

        let dim = Style::default().add_modifier(Modifier::DIM);
        let title_style = Style::default()
            .fg(Color::Rgb(0x51, 0xAF, 0xEF))
            .add_modifier(Modifier::BOLD);
        let selected_style = Style::default().add_modifier(Modifier::REVERSED | Modifier::BOLD);
        let dir_style = Style::default().fg(Color::Rgb(0x51, 0xAF, 0xEF));
        let sentinel_style = Style::default()
            .fg(Color::Rgb(0x98, 0xC3, 0x79))
            .add_modifier(Modifier::BOLD);

        let mut y = area.y;

        // Top border
        if y < area.y + area.height {
            let sep = "\u{2500}".repeat(area.width as usize);
            buf.set_line(area.x, y, &Line::from(Span::styled(sep, dim)), area.width);
            y += 1;
        }

        // Title
        if y < area.y + area.height {
            buf.set_line(
                area.x,
                y,
                &Line::from(Span::styled(&self.title, title_style)),
                area.width,
            );
            y += 1;
        }

        // Input value
        if y < area.y + area.height {
            buf.set_line(
                area.x,
                y,
                &Line::from(Span::raw(&self.value)),
                area.width,
            );
            y += 1;
        }

        if self.mode == InputMode::Path {
            // Separator
            if y < area.y + area.height {
                let sep = "\u{2500}".repeat(area.width as usize);
                buf.set_line(area.x, y, &Line::from(Span::styled(sep, dim)), area.width);
                y += 1;
            }

            let max_visible = self.max_visible_entries();
            let end = (self.dir_scroll + max_visible).min(self.dir_entries.len());

            for i in self.dir_scroll..end {
                if y >= area.y + area.height {
                    break;
                }
                let entry = &self.dir_entries[i];
                let text = match entry.kind {
                    DirEntryKind::Sentinel => format!("  \u{25b8} {}", entry.name),
                    DirEntryKind::Dir => format!("  {}/", entry.name),
                    DirEntryKind::File => format!("  {}", entry.name),
                };

                let style = if i == self.dir_cursor {
                    selected_style
                } else {
                    match entry.kind {
                        DirEntryKind::Sentinel => sentinel_style,
                        DirEntryKind::Dir => dir_style,
                        DirEntryKind::File => dim,
                    }
                };

                buf.set_line(area.x, y, &Line::from(Span::styled(text, style)), area.width);
                y += 1;
            }

            if self.dir_entries.is_empty() && y < area.y + area.height {
                buf.set_line(
                    area.x,
                    y,
                    &Line::from(Span::styled("  (empty)", dim)),
                    area.width,
                );
                y += 1;
            }
        }

        // Help line at the bottom
        let help_y = area.y + area.height - 1;
        if help_y > y || y >= area.y + area.height {
            let help = if self.mode == InputMode::Path {
                " esc cancel  tab complete  \u{2191}\u{2193} select  enter open/confirm"
            } else {
                " esc cancel"
            };
            buf.set_line(
                area.x,
                help_y,
                &Line::from(Span::styled(help, dim)),
                area.width,
            );
        }
    }

    fn split_dir_prefix<'a>(path: &'a str) -> (&'a str, &'a str) {
        if path.ends_with('/') {
            (path, "")
        } else {
            match path.rfind('/') {
                Some(i) => (&path[..=i], &path[i + 1..]),
                None => (path, ""),
            }
        }
    }

    fn update_dir_listing(&mut self) {
        let path = self.expand_path(&self.value);
        let (dir, prefix) = Self::split_dir_prefix(&path);
        let prefix_lower = prefix.to_lowercase();

        let mut real_entries: Vec<DirEntry> = match fs::read_dir(dir) {
            Ok(entries) => entries
                .filter_map(|e| e.ok())
                .filter(|e| {
                    let name = e.file_name().to_string_lossy().to_string();
                    !name.starts_with('.')
                        && (prefix_lower.is_empty()
                            || name.to_lowercase().starts_with(&prefix_lower))
                })
                .map(|e| {
                    let name = e.file_name().to_string_lossy().to_string();
                    let kind = if e.file_type().map(|ft| ft.is_dir()).unwrap_or(false) {
                        DirEntryKind::Dir
                    } else {
                        DirEntryKind::File
                    };
                    DirEntry { name, kind }
                })
                .collect(),
            Err(_) => Vec::new(),
        };

        // Sort real entries: dirs first, then alphabetically
        real_entries.sort_by(|a, b| {
            let a_dir = a.kind == DirEntryKind::Dir;
            let b_dir = b.kind == DirEntryKind::Dir;
            b_dir.cmp(&a_dir).then(a.name.cmp(&b.name))
        });

        // Pin a synthetic "Select folder" row at the top of every listing so users
        // have a clear affordance for selecting the current directory.
        self.dir_entries = Vec::with_capacity(real_entries.len() + 1);
        self.dir_entries.push(DirEntry {
            name: "Select folder".to_string(),
            kind: DirEntryKind::Sentinel,
        });
        self.dir_entries.extend(real_entries);

        if self.dir_cursor >= self.dir_entries.len() {
            self.dir_cursor = self.dir_entries.len().saturating_sub(1);
        }
    }

    fn tab_complete(&mut self) {
        let path = self.expand_path(&self.value);
        let (dir, prefix) = Self::split_dir_prefix(&path);
        let prefix_lower = prefix.to_lowercase();

        let matches: Vec<String> = match fs::read_dir(dir) {
            Ok(entries) => entries
                .filter_map(|e| e.ok())
                .filter(|e| {
                    e.file_type().map(|ft| ft.is_dir()).unwrap_or(false)
                        && e.file_name()
                            .to_string_lossy()
                            .to_lowercase()
                            .starts_with(&prefix_lower)
                })
                .map(|e| e.file_name().to_string_lossy().to_string())
                .collect(),
            Err(_) => Vec::new(),
        };

        if matches.len() == 1 {
            self.value = format!("{}{}/", dir, matches[0]);
            self.cursor_pos = self.value.len();
        } else if matches.len() > 1 {
            // Find common prefix
            let mut common = matches[0].clone();
            for m in &matches[1..] {
                let shared: String = common
                    .chars()
                    .zip(m.chars())
                    .take_while(|(a, b)| a == b)
                    .map(|(a, _)| a)
                    .collect();
                common = shared;
            }
            if common.len() > prefix.len() {
                self.value = format!("{}{}", dir, common);
                self.cursor_pos = self.value.len();
            }
        }

        self.dir_cursor = 0;
        self.dir_scroll = 0;
        self.update_dir_listing();
        if self.dir_entries.len() > 1 {
            self.dir_cursor = 1;
        }
    }

    fn expand_path(&self, path: &str) -> String {
        crate::app::expand_tilde(path)
    }

    fn max_visible_entries(&self) -> usize {
        // Reserve lines for: border, title, input, separator, help
        let available = self.height as usize;
        available.saturating_sub(5).max(3)
    }
}
